pkg/registrar/registrar.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package registrar
+
+import (
+	"fmt"
+	"sync"
+)
+
+// TODO(random-liu): Add unit test.
+
+// Registrar stores name to key, and key to name indexes.
+// Names must be unique.
+// Registrar is safe for concurrent access.
+type Registrar struct {
+	sync.Mutex
+	nameToKey map[string]string
+	keyToName map[string]string
+}
+
+// NewRegistrar creates a new Registrar with the an empty index.
+func NewRegistrar() *Registrar {
+	return &Registrar{
+		nameToKey: make(map[string]string),
+		keyToName: make(map[string]string),
+	}
+}
+
+// Reserve registers a key to a name.
+// Reserve is idempotent.
+// Attempting to reserve a key to a name that already exists results in an error.
```

Comment:
says returns in error ... but below it returns nil if the name key pair already exists..

probably should say can only assign one name to key pair.. 

Should also say both name and key must be unique..